org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseDocumentLifeCycleHandler.java

```diff
@@ -329,7 +334,13 @@ public IProblemRequestor getProblemRequestor(ICompilationUnit workingCopy) {
 		};
 		int flags = ICompilationUnit.FORCE_PROBLEM_DETECTION | ICompilationUnit.ENABLE_BINDINGS_RECOVERY | ICompilationUnit.ENABLE_STATEMENTS_RECOVERY;
 		synchronized(reconcileLock) {
-			unit.reconcile(ICompilationUnit.NO_AST, flags, wcOwner, monitor);
+			if (!org.eclipse.jdt.internal.core.CompilationUnit.DOM_BASED_OPERATIONS) {
+				unit.reconcile(ICompilationUnit.NO_AST, flags, wcOwner, monitor);
+			} else {
+				// ReconcileWorkingCopyOperation.executeOperation() creates AST twice with NO_AST
```

Comment:
Should the code be changed upstream so the dom-based approach supports `NO_AST` ?

Reply:
cc @mickaelistria @datho7561  @robstryker @fbricon @testforstephen

Reply:
It seems like in DOM-based-operations, we are always creating the AST using ASTParser in order to collect the problems. This implementation is decoupled from any classes internal to ecj or javac, but it means we always create the AST, and can't skip steps that would make it faster.

I think it's possible to change our code in `ReconcileWorkingCopyOperation` to avoid this:

Add another method to `ICompilationUnitResolver` that has a similar interface to `CompilationUnitProblemFinder.process()` that will allow us to collect problems without creating the AST. We would need to write implementations for ECJ and javac. Then, we could rewrite `ReconcileWorkingCopyOperation` to get the `ICompilationUnitResolver` that's configured and call our new method, so we don't need to check the DOM_BASED_OPERATIONS flag in `ReconcileWorkingCopyOperation`. This still avoids referencing ecj internals, since `ICompilationUnitResolver` is outside ECJ.

@mickaelistria what do you think?

Reply:
> I think it's possible to change our code in ReconcileWorkingCopyOperation to avoid this:

Could we apply the following patch:
```
diff --git a/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/ReconcileWorkingCopyOperation.java b/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/ReconcileWorkingCopyOperation.java
index 71772461d9..f7cb62aaec 100644
--- a/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/ReconcileWorkingCopyOperation.java
+++ b/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/ReconcileWorkingCopyOperation.java
@@ -218,9 +218,9 @@ public class ReconcileWorkingCopyOperation extends JavaModelOperation {
                                                for (Entry<String, List<CategorizedProblem>> entry : groupedProblems.entrySet()) {
                                                        this.problems.put(entry.getKey(), entry.getValue().toArray(CategorizedProblem[]::new));
                                                }
-                                               if (this.astLevel != ICompilationUnit.NO_AST) {
+                                               // if (this.astLevel != ICompilationUnit.NO_AST) {
                                                        this.ast = newAST;
-                                               }
+                                               // }
                                        } catch (AbortCompilationUnit ex) {
                                                var problem = ex.problem;
                                                if (problem == null && ex.exception instanceof IOException ioEx) {

```

Reply:
In any case, this condition shouldn't make it into JDT-LS, the same code should work without checking whether it's DOM or ECJ. So let's just remove workarounds.
If there is an issue with DOM based stuff, please report it to the jdt-incubator repo.
Building the AST does cost much extra CPU (it's neglectable compared to the cost of eg resolving symbols to actual .class files in jars), but keeping it may retain some extra memory. However, I don't think you're in this case and thus the code should just remain the same as with ECJ.

Reply:
> In any case, this condition shouldn't make it into JDT-LS, the same code should work without checking whether it's DOM or ECJ. So let's just remove workarounds.

I have removed it.

Reply:
> Would that help? 

Yes, it would. We should remember AST if we create it.

The AST will still be built. I think it would be better to avoid building the AST, right?

>  I think it would be better to avoid building the AST, right?

Right. ecj works in that way.

> Here is my idea implemented:
https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/947

I have tried to apply your PR. When I call
```
unit.reconcile(ICompilationUnit.NO_AST, flags, wcOwner, monitor);
```
in Java LS, I get the following error:
```
[Error - 7:49:39 PM] Nov 13, 2024, 7:49:39?PM An internal error occurred during: "Publish Diagnostics".
error: no source files
java.lang.IllegalStateException: error: no source files
	at jdk.compiler/com.sun.tools.javac.main.Arguments.reportDiag(Arguments.java:908)
	at jdk.compiler/com.sun.tools.javac.main.Arguments.validate(Arguments.java:466)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.prepareCompiler(JavacTaskImpl.java:194)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.parseInternal(JavacTaskImpl.java:257)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.invocationHelper(JavacTaskImpl.java:152)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.parse(JavacTaskImpl.java:248)
	at org.eclipse.jdt.core.dom.JavacCompilationUnitResolver.findProblems(JavacCompilationUnitResolver.java:1269)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.makeConsistent(ReconcileWorkingCopyOperation.java:193)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:98)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:751)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:816)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1439)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDocumentLifeCycleHandler.publishDiagnostics(BaseDocumentLifeCycleHandler.java:336)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDocumentLifeCycleHandler.publishDiagnostics(BaseDocumentLifeCycleHandler.java:299)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDocumentLifeCycleHandler$PublishDiagnosticJob.run(BaseDocumentLifeCycleHandler.java:781)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

Reply:
Thank you, I'll take a look into it

Reply:
I pushed a change that might fix it, but it seems none of the test cases or manual testing I've been doing actually uses this method, so I have no idea if it works. I've got to step out for a bit but when I'm back I'll look into running it in vscode-java.

Reply:
It seems to be working now from my testing

Reply:
Where is it?

Reply:
https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/compare/29fee8521b250dc94095e648f7da9ec323fa59eb..7f0b7b0c5366bf01822e8334562cc7779a100daf I force pushed a change to that PR

Reply:
I have tested https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/947/commits/7f0b7b0c5366bf01822e8334562cc7779a100daf Please see my previous comment.

Reply:
Thanks. I wasn't testing it locally correctly, my bad

Reply:
Okay, I figured out how to test it locally correctly, and addressed that bug, it should be good now

Reply:
> https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/compare/29fee8521b250dc94095e648f7da9ec323fa59eb..7f0b7b0c5366bf01822e8334562cc7779a100daf I force pushed a change to that PR

I have tried https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/947/commits/6cee4e3effd470589fd7337be242ecad6a5b828f 
Eclipse
![eclipse947](https://github.com/user-attachments/assets/3db9188a-5ade-49cf-9aeb-b1495e0f9bdb)
VS Code
![vscode947](https://github.com/user-attachments/assets/64abd390-807e-4fc1-8143-f66ee24f3c38)

I can also see the following exception in VS Code log
```
[Error - 5:52:39 PM] Nov 14, 2024, 5:52:39?PM An internal error occurred during: "Publish Diagnostics".
null
java.lang.IndexOutOfBoundsException
	at org.eclipse.jdt.ls.core.internal.DocumentAdapter.getChar(DocumentAdapter.java:207)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDiagnosticsHandler.getSourceStart(BaseDiagnosticsHandler.java:400)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDiagnosticsHandler.convertRange(BaseDiagnosticsHandler.java:363)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDiagnosticsHandler.toDiagnosticsArray(BaseDiagnosticsHandler.java:257)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDiagnosticsHandler.endReporting(BaseDiagnosticsHandler.java:148)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.reportProblems(ReconcileWorkingCopyOperation.java:147)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:116)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:751)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:816)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1439)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDocumentLifeCycleHandler.publishDiagnostics(BaseDocumentLifeCycleHandler.java:336)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDocumentLifeCycleHandler.publishDiagnostics(BaseDocumentLifeCycleHandler.java:299)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDocumentLifeCycleHandler$PublishDiagnosticJob.run(BaseDocumentLifeCycleHandler.java:781)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

Reply:
Can you please elaborate what problem is caused by the current code? I'm not sure we really need a change in the Javac-based reconciling to make your PR work as expected. What am I missing?

Reply:
You don't need to do anything.